### PR TITLE
Added an insertbefore option.

### PR DESCRIPTION
**Pull request by dorfsmay**
insertbeore option for the lineinfile module, in order to be able to insert a line **above** a given regex, rather and undernieth.

### CODE REVIEW

**Comment by mpdehaan** [comment]
Can you show me an example of this in a playbook?

**Comment by dorfsmay** [comment]
On 2013-02-11 06:07, Michael DeHaan wrote:

> Can you show me an example of this in a playbook?

```
 - name: Add JRE to Catalina.sh
   lineinfile: dest=${tomcat_dir}/bin/catalina.sh insertbefore='^# OS 
```

specific support.' regexp='^JRE_HOME=' line="JRE_HOME=${java_dir}" backup=yes

catalina.sh is 500 lines, I really don't want to template it, just to add a line!

## 

Yves.                                                  http://www.SollerS.ca/
                                  Unix/Linux and Python specialist in Calgary.
                                                        http://blog.zioup.org/

**Comment by dhozac** [comment]
This should use mutually_exclusive in module_common for insertafter and insertbefore.

False really doesn't make sense as a default value there, just remove that entirely and rely on it being None.

This breaks the case of not specifying insertafter/insertbefore at all, as insertafter no longer defaults to EOF. It should.

The logic for compiling the regexp is hard to follow.

```
if insertafter is not None and insertafter not in ('BOF', 'EOF'):
    insre = re.compile(insertafter)
elif insertbefore is not None and insertbefore not in ('BOF'):
    insre = re.compile(insertbefore)
else:
    insre = None
```

is far more readable to me.

Line 155 can be removed entirely, backwards compat consisting of a single part of an if-statement won't be removed.

default: is listed twice in the documentation string. In addition, it claims that the default is now to put the line first, and last in the file, which obviously isn't true.

**Comment by dorfsmay** [comment]
On 2013-02-11 16:03, Daniel Hokka Zakrisson wrote:

> This breaks the case of not specifying insertafter/insertbefore at all, as
> insertafter no longer defaults to EOF. It should.

Just to be 100% clear here, you mean insertafter should still default to EOF. 
Right?

**Comment by dhozac** [comment]
Yes.

**Comment by mpdehaan** [comment]
I'm ok with the conceptual concept of insertbefore.   Haven't had time to look at code yet.

**Comment by dorfsmay** [comment]
On 2013-02-12 18:47, Michael DeHaan wrote:

> I'm ok with the conceptual concept of insertbefore. Haven't had time to look
> at code yet.

Don't look at it yet...

I am modifying the existing code following Daniel's suggestion, and will 
re-submit a pull request later tonight.

## 

Yves.                                                  http://www.SollerS.ca/
                                  Unix/Linux and Python specialist in Calgary.
                                                        http://blog.zioup.org/

**Comment by dorfsmay** [comment]
Modified to follow Daniel Hokka Zakrisson's recommendations.

**Comment by mpdehaan** [comment]
merged in thanks, IMHO we need better tests for this module if someone wants to add them.  I didn't see any additions for insertbefore.

**Comment by dorfsmay** [comment]
On 2013-02-16 10:48, Michael DeHaan wrote:

> merged in thanks, IMHO we need better tests for this module if someone wants
> to add them. I didn't see any additions for insertbefore.

Yes, it is my intention to write a test for this. I'll also write some docs on 
how to write/run the tests for new modules, I wasn't able to figure it out 
until somebody on IRC helped me out.

## 

Yves.                                                  http://www.SollerS.ca/
                                  Unix/Linux and Python specialist in Calgary.
                                                        http://blog.zioup.org/

**Comment by mpdehaan** [comment]
Good deal, thanks.

This would be useful for some modules that are starting to get more
intelligent, especially where they don't need to do major changes to the OS
to test them out.

Test instructions could go on the module dev page.

Thanks!
